s3/src/main/scala/akka/stream/alpakka/s3/impl/MergeOrderedN.scala

```diff
@@ -0,0 +1,128 @@
+package akka.stream.alpakka.s3.impl
+
+import akka.annotation.InternalApi
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.stream.{Attributes, Inlet, Outlet, UniformFanInShape}
+
+import scala.collection.{immutable, mutable}
+
+@InternalApi private[impl] object MergeOrderedN {
+  /** @see [[MergeOrderedN]] */
+  def apply[T](inputPorts: Int, breadth: Int) =
+    new MergeOrderedN[T](inputPorts, breadth)
+}
+
+/**
+  * Takes multiple streams (in ascending order of input ports) whose elements will be pushed only if all elements from the
+  * previous stream(s) are already pushed downstream.
+  *
+  * The `breadth` controls how many upstream are pulled in parallel.
+  * That means elements might be received in any order, but will be buffered (if necessary) until their time comes.
+  *
+  * '''Emits when''' the next element from upstream (in ascending order of input ports) is available
+  *
+  * '''Backpressures when''' downstream backpressures
+  *
+  * '''Completes when''' all upstreams complete and there are no more buffered elements
+  *
+  * '''Cancels when''' downstream cancels
+  */
+@InternalApi private[impl] final class MergeOrderedN[T](val inputPorts: Int, val breadth: Int) extends GraphStage[UniformFanInShape[T, T]] {
```

Comment:
`ConcatPreFetch` or something like that would be more correct , merge in Akka streams generally mean emit in any order.